### PR TITLE
[batch 3] migrate 18 example notebooks V2 -> V3

### PR DESCRIPTION
**Pull request by lucasjia-aws**
## Migrate 18 example notebooks from SageMaker Python SDK V2 to V3

This PR adds V3 versions of 18 notebooks under `build_and_train_models/` and `deploy_and_monitor/`. Each V3 notebook is added as a `-v3.ipynb` sibling next to the original V2 notebook (V2 files are left untouched for reference). The migrations replace the V2 `Estimator`/framework-estimator + `Model`/`Predictor` workflow with the V3 `sagemaker-core` resource classes and, where a real training entry point is involved, `sagemaker.train.ModelTrainer`. Endpoint autoscaling, CloudWatch, deployment-guardrail, and Application Auto Scaling cells are raw boto3 (SDK-agnostic) and are kept unchanged.

### What changed per notebook

| Notebook | What it demonstrates | Key V2 → V3 changes |
| --- | --- | --- |
| `sm-scikit_build_your_own_container-v3.ipynb` | BYOC scikit decision-tree custom container (train + real-time endpoint + batch transform) | `Estimator.fit()` → `ModelTrainer(training_image=ECR image)` + `Compute`/`OutputDataConfig`; `tree.deploy()` → core `Model`/`EndpointConfig`/`Endpoint`; `tree.transformer().transform()` → core `Model`/`TransformJob`; `predictor.predict` → `Endpoint.invoke`. Container `Dockerfile` modernized ubuntu:18.04 + pinned 2019 packages → `python:3.10-slim-bullseye` so it builds/serves |
| `sm-script_mode_distributed_training_horovod_tensorflow-v3.ipynb` | Horovod (MPI) distributed TensorFlow script-mode training on MNIST | `TensorFlow(distributions={"mpi":...})` → `ModelTrainer(distributed=MPI(...))`; `entry_point`/`source_dir` → `SourceCode`; framework image → `image_uris.retrieve`; `fit({train,test})` → `train([InputData])`; session/role → `sagemaker.core.helper.session_helper` |
| `sm-tabtransformer_tabular_classification-v3.ipynb` | JumpStart TabTransformer tabular classification with Automatic Model Tuning (HPO) | Generic `Estimator` → `ModelTrainer.from_jumpstart_config(JumpStartConfig)`; `sagemaker.tuner.HyperparameterTuner` → `sagemaker.train.tuner.HyperparameterTuner` + `tuner.tune`; parameter ranges → `sagemaker.core.parameter`; best-job deploy via core `Model`/`EndpointConfig`/`Endpoint` |
| `estimator-pytorch-cpu-v3.ipynb` (sm-training-queues-pytorch) | AWS Batch `TrainingQueue` job-queuing demo (CPU) | PyTorch `Estimator` → `sagemaker.train.ModelTrainer`; `TrainingInput` → `InputData`; `max_run`/`output_path`/`entry_point` → `StoppingCondition`/`OutputDataConfig`/`SourceCode`; `sagemaker.aws_batch.TrainingQueue` → `sagemaker.train.aws_batch.TrainingQueue`; shared `queue_utils.py` switched to a V3/V2 try-except dual-path import |
| `estimator-pytorch-gpu-v3.ipynb` (sm-training-queues-pytorch) | AWS Batch `TrainingQueue` job-queuing demo (GPU, FSDP+QLoRA fine-tune) | Same as CPU variant; `distribution={"torch_distributed"}` → `distributed=Torchrun()`; `sagemaker.s3.S3Uploader` → `sagemaker.core.s3.S3Uploader`; `sagemaker.config` → `sagemaker.core.config` |
| `sm-a_b_testing-v3.ipynb` | A/B testing: two XGBoost models as two production variants on one endpoint, weight-based traffic shifting | `sm_session.create_model` → `Model.create(ContainerDefinition)`; `production_variant` dict → `ProductionVariant` shape; `endpoint_from_production_variants` → `EndpointConfig`/`Endpoint.create`; `invoke_endpoint(TargetVariant=)` → `Endpoint.invoke(target_variant=)`; `update_endpoint_weights_and_capacities` → `Endpoint.update_weights_and_capacities([DesiredWeightAndCapacity])` |
| `sm-app_autoscaling_realtime_endpoints-v3.ipynb` | Faster Application Auto Scaling (variant-level, high-resolution metric) on a Llama-3-8B TGI endpoint | `get_huggingface_llm_image_uri` → `image_uris.retrieve(framework="huggingface-llm", ...)`; `HuggingFaceModel(env=config).deploy()` → core `Model`/`EndpointConfig(ProductionVariant)`/`Endpoint`; `predictor.predict(messages)` → `Endpoint.invoke`. Application Auto Scaling / llmperf cells unchanged (boto3) |
| `sm-app_autoscaling_realtime_endpoints_inference_components-v3.ipynb` | Faster autoscaling via Inference Components + high-resolution Application Auto Scaling on a Llama-3-8B TGI IC | boto3 `create_endpoint_config`/`create_endpoint`/`create_model` → core `EndpointConfig`/`Endpoint`/`Model`; `create_inference_component` + poll → `InferenceComponent.create` + `wait_for_status`; `Predictor(component_name).predict` → `Endpoint.invoke(inference_component_name=...)` |
| `sm-app_autoscaling_realtime_endpoints_step_scaling-v3.ipynb` | Same as autoscaling, using Step Scaling (manual policy + `put_metric_alarm`) | Same deploy/inference migration as the target-tracking twin; step-scaling + CloudWatch alarm cells unchanged (boto3) |
| `sm-async_inference_walkthrough-v3.ipynb` | Async inference walkthrough (boto3) — XGBoost async endpoint + S3 payloads | `import sagemaker` + `sagemaker.session.Session()`/`get_execution_role()` → `sagemaker.core.helper.session_helper`; `from sagemaker import image_uris` → `sagemaker.core.image_uris`. `invoke_endpoint_async` + Application Auto Scaling kept on boto3 |
| `sm-async_inference_with_python_sdk-v3.ipynb` | Async inference via the SageMaker Python SDK (sibling of the boto3 version) | `Model(predictor_cls=Predictor)` → `ModelBuilder(image_uri, model_path)` + `build()`/`deploy(inference_config=AsyncInferenceConfig)`; `AsyncInferenceConfig` → `sagemaker.core.inference_config`; `AsyncPredictor.predict_async().output_path` → `Endpoint.invoke_async().output_location` |
| `sm-batch_inference_with_torchserve-v3.ipynb` | TorchServe real-time dynamic-batching (BERT seq-classification, script mode) | `PyTorchModel(source_dir, entry_point, env).deploy()` → `repack_model` (bundles `code/` into the model tar) + core `Model`/`EndpointConfig`/`Endpoint` with `ContainerDefinition(environment=...)` carrying the `SAGEMAKER_TS_*` batching vars; `Predictor.predict` → `Endpoint.invoke` |
| `sm-deployment_guardrails_update_inference_endpoint_with_linear_traffic_shifting-v3.ipynb` | Deployment guardrails — linear traffic shifting (rollback + success) | `from sagemaker import get_execution_role, session` → `sagemaker.core.helper.session_helper`; `from sagemaker.s3 import S3Uploader` → `sagemaker.core.s3.S3Uploader`; `from sagemaker import image_uris` → `sagemaker.core.image_uris`. `update_endpoint` with `BlueGreenUpdatePolicy` LINEAR + CloudWatch alarms unchanged (boto3) |
| `sm-deployment_guardrails_update_inference_endpoint_with_rolling_deployment-v3.ipynb` | Deployment guardrails — rolling deployment (rollback + success) | Same session/S3/image_uris migration as linear; `update_endpoint` with `RollingUpdatePolicy` + `AutoRollbackConfiguration` unchanged (boto3) |
| `sm-deployment_guardrails_update_inference_endpoint_with_with_canary_traffic_shifting-v3.ipynb` | Deployment guardrails — canary traffic shifting (rollback + success) | Same session/S3/image_uris migration; `update_endpoint` with `BlueGreenUpdatePolicy` CANARY + `AutoRollbackConfiguration` unchanged (boto3) |
| `sm-marketplace_using_model_package_arn-v3.ipynb` | Consume a Marketplace model-package ARN for batch transform + a live endpoint | `ModelPackage(model_package_arn=...)` → `Model.create(ContainerDefinition(model_package_name=arn), enable_network_isolation=True)`; `model.transformer()/.transform()` → `TransformJob.create`; `model.deploy()` → core `EndpointConfig`/`Endpoint`; `Predictor.predict` → `Endpoint.invoke` |
| `sm-mme_with_torchserve-v3.ipynb` | SageMaker Multi-Model Endpoint (MME) hosting multiple TorchServe models | `MultiDataModel(model=Model(...))` → `ContainerDefinition(mode="MultiModel", multi_model_config=MultiModelConfig(), model_data_url=)` + core `Model`/`EndpointConfig`/`Endpoint`; `mme.add_model()` → `S3Uploader.upload`; `Predictor.predict(target_model=...)` → `Endpoint.invoke(target_model=...)` |
| `mnist-train-using-pytorch-v3.ipynb` (sm-multi_cloud_deployment_with_onnx) | Multi-cloud demo: train PyTorch MNIST on SageMaker, export to ONNX (only the training stage touches the SDK) | `sagemaker.pytorch.PyTorch` → `image_uris.retrieve(framework="pytorch", ...)` + `ModelTrainer(training_image, SourceCode, Compute, OutputDataConfig, StoppingCondition)`; `estimator.fit({training,testing})` → `train([InputData])`; `estimator.model_data` → `model_trainer._latest_training_job.model_artifacts.s3_model_artifacts` |

### Common migration themes

- **Session/role:** `import sagemaker` + `sagemaker.Session()` / `sagemaker.get_execution_role()` → `from sagemaker.core.helper.session_helper import Session, get_execution_role`.
- **Image lookup:** `sagemaker.image_uris` → `sagemaker.core.image_uris`.
- **Training/deploy:** the V2 `Estimator`/framework-estimator `.fit()`/`.deploy()` flow is replaced either by `sagemaker-core` resource objects (`TrainingJob`, `Model`, `EndpointConfig`, `Endpoint`, `InferenceComponent`, `TransformJob`) or by `sagemaker.train.ModelTrainer` (+ `SourceCode`/`Compute`/`InputData` configs) where a custom/framework training entry point is used.
- **Inference:** `Predictor.predict` → `Endpoint.invoke` (real-time), `Endpoint.invoke_async` (async), or `Endpoint.invoke(target_variant=/target_model=/inference_component_name=)`.
- **Unchanged:** Application Auto Scaling, CloudWatch, deployment-guardrail `update_endpoint`, and Ground-Truth-style boto3 calls are SDK-agnostic and left as-is.

### Non-notebook changes in this PR

- `build_and_train_models/sm-scikit_build_your_own_container/container/Dockerfile`: modernized Python 2 / ubuntu:18.04 + 2019-pinned packages → `python:3.10-slim-bullseye` with compatible packages (train/serve logic unchanged).
- `build_and_train_models/sm-training-queues-pytorch/smtj_batch_utils/queue_utils.py`: V3/V2 try-except dual-path import (`sagemaker.train.aws_batch` first, `sagemaker.aws_batch` fallback) so both migrated and unmigrated sibling notebooks work.
- `deploy_and_monitor/sm-app_autoscaling_realtime_endpoints/utils/llmperf.py` and `.../sm-app_autoscaling_realtime_endpoints_step_scaling/utils/llmperf.py`: re-enabled `os.chmod(script_path, 0o755)` — `subprocess.Popen([script])` needs the execute bit, which git does not always preserve, so it is set explicitly before launching `trigger_autoscaling.sh`.

### Notes / caveats

- **`sm-app_autoscaling_realtime_endpoints_step_scaling-v3.ipynb`** is migrated and functionally validated (endpoint InService, inference OK, step-scaling policies + alarms created, llmperf-driven scale-out observed) but was **not captured in a single all-green executed run** due to intermittent AWS `InsufficientInstanceCapacity` for the GPU endpoint (capacity, not code/quota).

## Merge Checklist

_Put an `x` in the boxes that apply._

- [x] I have verified that my PR does not contain any new notebook/s which demonstrate a SageMaker functionality already showcased by another existing notebook in the repository
- [x] I have read the [CONTRIBUTING](https://github.com/aws/amazon-sagemaker-examples/blob/default/CONTRIBUTING.md) doc and adhered to the guidelines regarding folder placement, notebook naming convention and example notebook best practices
- [x] I have updated the necessary documentation, including the README of the appropriate folder as well as the index.rst file
- [x] I have tested my notebook(s) and ensured it runs end-to-end
- [ ] I have linted my notebook(s) and code using `python3 -m black -l 100 {path}/{notebook-name}.ipynb`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File:      deploy_and_monitor/sm-app_autoscaling_realtime_endpoints/utils/llmperf.py [hunk 1]
```diff
@@ -35,7 +35,7 @@ def trigger_auto_scaling(creds, region, endpoint_name, num_concurrent_requests):
         print(f"LLMPerf shell script file not found at {script_path}")
 
     # Make sure the script is executable
-    # os.chmod(script_path, 0o755)
+    os.chmod(script_path, 0o755)
```
[the remark this answers is not in the document]
Comment (lucasjia-aws) [reply      deploy_and_monitor/sm-app_autoscaling_realtime_endpoints/utils/llmperf.py]: fixed with 0o700

File:      deploy_and_monitor/sm-app_autoscaling_realtime_endpoints_step_scaling/utils/llmperf.py [hunk 2]
```diff
@@ -34,8 +34,9 @@ def trigger_auto_scaling(creds, region, endpoint_name, num_concurrent_requests):
     else:
         print(f"LLMPerf shell script file not found at {script_path}")
 
-    # Make sure the script is executable
-    # os.chmod(script_path, 0o755)
+    # Make sure the script is executable (git does not always preserve the
+    # execute bit, so set it explicitly before launching the subprocess).
+    os.chmod(script_path, 0o755)
```
[the remark this answers is not in the document]
Comment (lucasjia-aws) [reply      deploy_and_monitor/sm-app_autoscaling_realtime_endpoints_step_scaling/utils/llmperf.py]: fixed with 0o700

**Review by jam-jee — APPROVED** [review]
